python/ray/dashboard/modules/job/job_head.py

```diff
@@ -213,6 +220,53 @@ async def choose_agent(self) -> Optional[JobAgentSubmissionClient]:
 
             return self._agents[node_id]
 
+    async def _get_head_node_agent(self) -> Optional[JobAgentSubmissionClient]:
+        """Retrieves HTTP client for `JobAgent` running on the Head node"""
+
+        head_node_id = await get_head_node_id(self._gcs_aio_client)
+
+        if not head_node_id:
+            logger.warning("Head node id has not yet been persisted in GCS")
+            return None
+
+        if head_node_id not in self._agents:
+            agent_infos = await self._fetch_agent_infos(target_node_ids=[head_node_id])
+            if head_node_id not in agent_infos:
+                logger.error("Head node agent's information was not found")
+                return None
+
+            agent_info = agent_infos[head_node_id]
+
+            node_ip = agent_info["ipAddress"]
+            http_port = agent_info["httpPort"]
+            agent_http_address = f"http://{node_ip}:{http_port}"
+
+            self._agents[head_node_id] = JobAgentSubmissionClient(agent_http_address)
+
+        return self._agents[head_node_id]
+
+    @staticmethod
+    async def _fetch_agent_infos(target_node_ids: Optional[List[str]] = None):
+        """Fetches agent infos for nodes identified by provided node-ids (for all
+        nodes if not provided)
+
+        NOTE: This call will block until there's at least 1 valid agent info populated
+        """
+
+        while True:
```

Comment:
shall we add a timeout here or at the call site? and/or at least an info-level log when we're sleeping

Reply:
Good questions:

 - So previous code would just loop indefinitely w/o any timeouts. We can add one but we will have to retry at that level then, since we can't proceed w/o agent-info
 - There's actually warning log inside `get_agent_infos` every time it misses to fetch requested data so it would be complaining in the log pretty vocally